File: levelledmobs-plugin-bukkit/src/main/java/me/lokka30/levelledmobs/bukkit/listeners/EntitySpawnListener.java
package me.lokka30.levelledmobs.bukkit.listeners;

import me.lokka30.levelledmobs.bukkit.LevelledMobs;
import me.lokka30.levelledmobs.bukkit.api.data.keys.EntityKeyStore;
import me.lokka30.levelledmobs.bukkit.data.InternalEntityDataUtil;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.entity.EntitySpawnEvent;

public final class EntitySpawnListener extends ListenerWrapper {

    public EntitySpawnListener() {
        super("org.bukkit.event.entity.EntitySpawnEvent", true);
    }

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    public void handle(final EntitySpawnEvent event) {
        /*
        LevelledMobs only concerns LivingEntities
         */
        if(!(event.getEntity() instanceof LivingEntity))
            return;

        final LivingEntity entity = (LivingEntity) event.getEntity();

        /*
        Check if the entity has any non-persistent metadata to migrate
         */
        // wasSummoned
        var wasSummonedKeyStr = EntityKeyStore.wasSummoned.toString();
        InternalEntityDataUtil.setWasSummoned(
            entity,
            entity.hasMetadata(wasSummonedKeyStr) &&
                entity.getMetadata(wasSummonedKeyStr).stream().anyMatch(val -> val.asInt() == 1)
        );
        entity.removeMetadata(wasSummonedKeyStr, LevelledMobs.getInstance());
    }
}

File: levelledmobs-plugin-bukkit/src/main/kotlin/me/lokka30/levelledmobs/bukkit/listeners/EntitySpawnListener.kt
package me.lokka30.levelledmobs.bukkit.listeners

import me.lokka30.levelledmobs.bukkit.LevelledMobs
import me.lokka30.levelledmobs.bukkit.api.data.keys.EntityKeyStore
import me.lokka30.levelledmobs.bukkit.data.InternalEntityDataUtil
import org.bukkit.entity.LivingEntity
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.entity.EntitySpawnEvent

/*
FIXME comment
 */
class EntitySpawnListener : ListenerWrapper(
    "org.bukkit.event.entity.EntitySpawnEvent"
) {

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun handle(event: EntitySpawnEvent) {
        /*
        LevelledMobs only concerns LivingEntities
         */
        if(event.entity !is LivingEntity) return

        /*
        Firstly, let's check if the entity has any non-persistent metadata to migrate.
         */
        // wasSummoned
        val wasSummonedKeyStr = EntityKeyStore.wasSummoned.toString()
        InternalEntityDataUtil.setWasSummoned(
            event.entity as LivingEntity,
            event.entity.hasMetadata(wasSummonedKeyStr) &&
                    event.entity.getMetadata(wasSummonedKeyStr).any{it.asInt() == 1}
        )
        event.entity.removeMetadata(wasSummonedKeyStr, LevelledMobs.instance!!)
    }

}